src/network/networkd-link.c

```diff
@@ -4119,100 +4150,50 @@ int link_save(Link *link) {
-                                        space = true;
-                }
+                /************************************************************/
 
                 if (link->network->dhcp6_use_ntp && dhcp6_lease) {
-                        const struct in6_addr *in6_addrs;
-                        char **hosts;
-
-                        r = sd_dhcp6_lease_get_ntp_addrs(dhcp6_lease,
-                                                         &in6_addrs);
-                        if (r > 0) {
-                                if (space)
-                                        fputc(' ', f);
-                                serialize_in6_addrs(f, in6_addrs, r);
-                                space = true;
-                        }
-
-                        r = sd_dhcp6_lease_get_ntp_fqdn(dhcp6_lease, &hosts);
-                        if (r > 0)
-                                fputstrv(f, hosts, NULL, &space);
+                        (void) sd_dhcp6_lease_get_ntp_addrs(dhcp6_lease, &in6_addrs);
+                        (void) sd_dhcp6_lease_get_ntp_fqdn(dhcp6_lease, &in6_hosts);
                 }
-
-                fputc('\n', f);
+                serialize_addresses(f, "NTP", NULL,
+                                    link->ntp ?: link->network->ntp,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_NTP_SERVERS,
+                                    link->network->dhcp_use_ntp,
+                                    in6_addrs, in6_hosts);
+
+                /************************************************************/
+
+                serialize_addresses(f, "SIP", NULL,
+                                    link->network->sip,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_SIP_SERVERS,
+                                    link->network->dhcp_use_sip,
+                                    NULL, NULL);
+
+                serialize_addresses(f, "POP3_SERVERS", NULL,
+                                    link->network->pop3,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_POP3_SERVERS,
+                                    true,
+                                    NULL, NULL);
+
+                serialize_addresses(f, "SMTP_SERVERS", NULL,
+                                    link->network->smtp,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_SMTP_SERVERS,
+                                    true,
+                                    NULL, NULL);
+
+                serialize_addresses(f, "LPR_SERVERS", NULL,
+                                    link->network->lpr,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_LPR_SERVERS,
+                                    true,
+                                    NULL, NULL);
```

Comment:
hmm, I think we should probably drop the _SERVERS suffix for this too, no? i.e. DNS and SIP and NTP don't have such a suffix, and these are just the same

Reply:
Yeah, I think we should do that, but that is mostly cosmetic. I would prefer to figure out the part I added to TODO first, i.e. how to actually test this. Apparently this wasn't tested at all, and various link files that I have on running machines only have a small subset of fields defined, so it's hard to test serialization comprehensively.